components/eamxx/src/share/io/scorpio_output.cpp

```diff
@@ -384,6 +382,43 @@ run (const std::string& filename,
     stop_timer("EAMxx::IO::horiz_remap");
   }
 
+  // Update all of the averaging count views (if needed)
+  // The strategy is as follows:
+  // For the update to the averaged value for this timestep we need to track if
+  // a point in a specific layout is "filled" or not.  So we create a set of local
+  // temporary views for each layout that are either 0 or 1 depending on if the
+  // value is filled or unfilled.
+  // We then use these values to update the overall average count views for that layout.
+  if (m_track_avg_cnt && m_add_time_dim) {
+    // Note, we assume that all fields that share a layout are also masked/filled in the same
+    // way.  If, we need to handle a case where only a subset of output variables are expected to
+    // be masked/filled then the recommendation is to request those variables in a separate output
+    // stream.
+    // We cycle through all fields and mark points that are filled/masked in the local views.  First
+    // initialize them to 1 representing unfilled.
+    for (const auto name : m_avg_cnt_names) {
+      auto& dev_view = m_local_tmp_avg_cnt_views_1d.at(name);
+      Kokkos::deep_copy(dev_view,1.0);
+    }
+    // Now we cycle through all the fields
+    for (const auto name : m_fields_names) {
+      auto field    = get_field(name,"io");
+      auto lookup   = m_field_to_avg_cnt_map.at(name);
+      auto dev_view = m_local_tmp_avg_cnt_views_1d.at(lookup);
+      update_avg_cnt_view(field,dev_view);
```

Comment:
Aren't you calling this fcn many times, if there are 2+ fields with the same avg count field?

Reply:
yes, the idea being that we want consistency over each layout.  So one variable may have a fill value, while others won't.  I think I outline in a comment that for such a case we will mask that value for all fields that share that layout.  Thus, I need to loop over all fields and check for fill values because I won't know ahead of time which one(s) have been filled.

essentially, if I have variable A and B on layout L and only A has a fill value at point `i`.  In this current implementation we would flag point `i` as "filled" and skip it for averaging calculations in both field `A` and field `B`.  In order to detect the fill value at point `i` I need to check both `A` and `B`.

I'm open to suggestions on how to approach this differently.  This was the only way I could think of that safely follows the assumption that all layouts are "filled" with the same pattern, and account for an edge case where some variable doesn't actually have a fill value.

Reply:
Well, if all fields with the same layout have the same mask, then you only need to process the lookup field once. The way it is now, I think you end up doing lookup+=N on the non-masked entries, where N is the number of fields that share that mask field...

Reply:
yes, my concern here is for cases where there are two fields with the same layout, but only 1 actually has the filled value.  Maybe the other has a reasonable value.  If I only check a single field for that layout then won't I be running the risk that I won't catch this case consistently.

I'm trying to anticipate future use cases that would be hard to debug.  If I only check one field and for some reason a user sets up a process that fills only a subset of variables on a specific layout then I run the risk of "averaging" over filled values.  I essentially don't trust that a in the future we won't have cases where the same layout see variables with and without fill values.  And there won't be an obvious error to throw in that case.

Reply:
If two fields have the same mask field attached to them, then they are masked the same way. If parametrization XYZ fills field "blah" only at even cols, then parametrization XYZ is responsible for attaching a different mask to "blah".

It's not a matter of "layout", but more a matter of "mask": if two fields are masked in the same way, then at every given entry they are either both valid or both invalid.

Reply:
Regardless of the motivation, with the code you have, if two fields have the same "lookup" field, you end up with avg_cnt+=N in some entries, which doesn't really make sense. The avg_cnt may even be different at different entries, no? What does the lookup field even represent here then?

Reply:
ahh yes, I see the confusion.  the lookup is a map of a `field` to the corresponding `mask` field for it.  So if N fields are on layout L then the lookup will have N pointing to 1 unique dev_view.  So the idea here is to cycle through every field associated with that dev_view and ensure that if that field has a masked value that the corresponding location in the mask view is represented as "masked".  The function "update_avg_cnt_view" might be a poor name for the function, if you look at it, all it does is set values to 0 if a mask is present.  So at the end you should only have a view of 0's and 1's.

Reply:
down below, the for loop there updates the true average count over the span of averaged time steps by adding the final view from this step.

Reply:
Oh, I see, then discard my +=N comment.

Nevertheless: you must be assuming that all field with the same mask have invalid values at the same location, right? Otherwise it would not make much sense: if field A is invalid where field B is valid and viceversa, the count will be 0 for both fields, even though each of them is valid in half of the entries...

Therefore, we should only have to call update_avg_cnt once for each mask field...

Reply:
Right, I agree, although w/ masked values things are a bit weird.  I'm trying to hedge against the case where field A has a valid value at a point where field B is invalid.  If we only check one field we have the potential of it being A, thus we will flag the value as valid then B gets an incorrect value in the averaging.  It shouldn't happen, but if it does it would a tricky bug to catch.

Reply:
The alternative would be to throw an error in this case.  The final alternative is to have each field get it's own masking array, but that will double the size of the restart history file, so I think that's a non-starter.

Reply:
I don't understand why field A would be valid if field B would be invalid, if they share the same mask field. If field A not only is remapped (like B) but also, say, masked out by radiation (to set it to invalid values over night columns), then field B should have a different mask attached to it.

So if two fields have the same mask field attached, they must have valid/invalid values in the _same_ locations.

Reply:
If we are talking only about vertical remapping, then yes, vertical remapping will apply the same mask to all fields on the same layout.  With this redundancy I'm attempting to tackle a more general application, if a specific atmospheric process ends up setting a value to the fill value for field A, because that is the only field it specifically cares about, we find ourselves in the case I described above.

TBH - I'm not excited about this approach, so I am open to push back or suggestions.  We could, for example, just support masking in the case of online remapping .  We don't currently have any processes that mask field values.  That being said, I don't think it is unreasonable to assume that in the future someone will introduce a process that wants to do masking.  For example, nudging may want to mask some nudging tendencies that are of bounds so we output nudging tendencies we find ourselves in this case.